### PR TITLE
C#: Add flow steps for View calls refering to Razor pages 

### PR DESCRIPTION
**Pull request by joefarebrother**
Adds a flow step from the `model` parameter of a `View` call to the Razor page it refers to using [this](https://learn.microsoft.com/en-us/aspnet/core/mvc/views/overview?view=aspnetcore-7.0#view-discovery) view discovery system, when this can be determined statically.

### CODE REVIEW

File: csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/gen_files.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# A script for generating code from .cshtml files, mimicking the output of the C# compiler with an option that is not available from the codeql test runner.
```
Comment (michaelnebel) [inline csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/gen_files.py:1]: How do I invoke this script properly? :-)

File: csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/Generated/Template.g [hunk 2]
```diff
@@ -0,0 +1,74 @@
+// A test file that mimics the output of compiling a `.cshtml` file
```
Comment (michaelnebel) [inline csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/Generated/Template.g:1]: How did you create this template?

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll [hunk 3]
```diff
@@ -0,0 +1,181 @@
+/** Provides definitions and flow steps related to Razor pages. */
+
+import csharp
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:3]:
```suggestion
private import csharp
```

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+private import codeql.util.FilePath
+private import semmle.code.csharp.frameworks.microsoft.AspNetCore
+
+/** A call to the `View` method */
+private class ViewCall extends MethodCall {
+  ViewCall() { this.getTarget().hasQualifiedName("Microsoft.AspNetCore.Mvc", "Controller", "View") }
+
+  /** Gets the `name` argument to this call, if any. */
+  string getNameArgument() {
+    exists(StringLiteral lit, int i | i in [0 .. 1] |
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:14]: Why is `i = 1` also included? According to the documentation of `Controller.View` this is the `masterName` (not sure about the semantics of this though).

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+      this.getTarget().getParameter(i).getType() instanceof StringType and
+      DataFlow::localExprFlow(lit, this.getArgument(i)) and
+      result = lit.getValue()
+    )
+  }
+
+  /** Gets the `model` argument to this call, if any. */
+  Expr getModelArgument() {
+    exists(int i | i in [0 .. 1] |
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:23]: There exist an overload of `View` where the model parameter is the third parameter - so maybe we should include `i = 2` in the range as well.

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+/** A call to the `View` method */
+private class ViewCall extends MethodCall {
+  ViewCall() { this.getTarget().hasQualifiedName("Microsoft.AspNetCore.Mvc", "Controller", "View") }
+
+  /** Gets the `name` argument to this call, if any. */
+  string getNameArgument() {
+    exists(StringLiteral lit, int i | i in [0 .. 1] |
+      this.getTarget().getParameter(i).getType() instanceof StringType and
+      DataFlow::localExprFlow(lit, this.getArgument(i)) and
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:16]: Is this to cater for?:
```csharp
var x = "my view path";
View(x);
```

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+  Method getActionMethod() {
+    result = this.getEnclosingCallable() and
+    result = this.getController().getAnActionMethod()
+  }
+
+  /**
+   * Gets the action name that this call refers to, if any.
+   * This is either the name argument, or the name of the action method calling this if there is no name argument.
+   */
+  string getActionName() {
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:39]: This string appears to reflect the "order" in "View Discovery".
There is also some mention of "Views" and "Views/Shared" which is also handled below. The documentation for view discovery doesn't mention anything about "Area": Do you know, if there exist some documentation for this as well?

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+      fp.hasViewCallWithIndex(vc, i) and
+      exists(RazorPage rp2 | ["", "~"] + rp2.getSourceFilepath() = fp.getNormalizedPath())
+    |
+      fp.getNormalizedPath() order by i
+    )
+}
+
+/** Gets the `i`th template for view discovery. */
+private string getViewSearchTemplate(int i, boolean isArea) {
+  i = 0 and result = "/Areas/{2}/Views/{1}/{0}.cshtml" and isArea = true
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:117]: I didn't find any mention of "Areas" in view discovery. Could you elaborate a bit?

File: csharp/ql/src/change-notes/2023-10-24-xss-flow-steps.md [hunk 9]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+Additional flow steps  to track flow from a `View` call in an MVC controller to the corresponding Razor page, which may result in additional results for queries such as `cs/web/xss`.
```
Comment (michaelnebel) [inline csharp/ql/src/change-notes/2023-10-24-xss-flow-steps.md:4]:
```suggestion
* Additional flow steps to track flow from a `View` call in an MVC controller to the corresponding Razor page, which may result in additional results for queries such as `cs/web/xss`.
```

**Review by michaelnebel — COMMENTED** [review]
Very nice work @joefarebrother !

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll (hunk 7 above)
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:39]: Areas are documented here: https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/areas?view=aspnetcore-7.0

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll (hunk 5 above)
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:23]: I can't find an overload with three or more parameters in the documentation?

File: csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/gen_files.py (hunk 1 above)
Comment (joefarebrother) [reply csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/gen_files.py:1]: What exceptions are you getting?

File: csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/Generated/Template.g (hunk 2 above)
Comment (joefarebrother) [reply csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/Generated/Template.g:1]: I took a generated file from the database from [this sample project](https://github.com/dsp-testing/csharp-f150), cut down to just the line containing the sink, and replaced anything dependent on the filepath with $PATHSLASH or $PATHUNDER.

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll (hunk 6 above)
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:16]: Yes

File: csharp/ql/test/resources/stubs/_frameworks/Microsoft.AspNetCore.App/Microsoft.AspNetCore.Razor.Runtime.cs [hunk 10]
```diff
@@ -22,6 +22,7 @@ public abstract class RazorCompiledItem
                     public abstract System.Collections.Generic.IReadOnlyList<object> Metadata { get; }
                     public abstract System.Type Type { get; }
                 }
+                [System.AttributeUsage(System.AttributeTargets.Assembly | System.AttributeTargets.Class, AllowMultiple=true, Inherited=false)]
```
Comment (hvitved) [inline csharp/ql/test/resources/stubs/_frameworks/Microsoft.AspNetCore.App/Microsoft.AspNetCore.Razor.Runtime.cs:25]: Now that https://github.com/github/codeql/pull/14609 has been merged, you can revert this change and rebase onto latest `main` instead.

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+
+  /**
+   * Gets the action name that this call refers to, if any.
+   * This is either the name argument, or the name of the action method calling this if there is no name argument.
+   */
+  string getActionName() {
+    result = this.getNameArgument()
+    or
+    not exists(this.getNameArgument()) and
+    result = this.getActionMethod().getName()
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:43]: Can action methods be generic? If so, we will probably need to strip away the `<,...,>` suffix.

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  /**
+   * Gets the filepath of the source file that this class was generated from,
+   * relative to the application root.
+   */
+  string getSourceFilepath() { result = attr.getArgument(2).(StringLiteral).getValue() }
+}
+
+private class ViewCallFlowStep extends TaintTracking::AdditionalTaintStep {
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:84]: Should this really be a value step and not a taint step?

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll (hunk 5 above)
Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:23]: Uh, I looked at the documentation for `[System.Web.Mvc](https://learn.microsoft.com/en-us/dotnet/api/system.web.mvc?view=aspnet-mvc-5.2)` - so please disregard.

**Review by tamasvajk — COMMENTED** [review]
Added some comments/questions mostly related to "MVC" vs "Razor Pages".

Also, a more general question:
Wouldn't we need to model flow from `View("path", model)` to an `ExecuteAsync` invocation of an instance of the page that has `"path"` after its `Model` property has been set to `model`?

File: csharp/ql/src/change-notes/2023-10-24-xss-flow-steps.md [hunk 13]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* Modelled additional flow steps to track flow from a `View` call in an MVC controller to the corresponding Razor page, which may result in additional results for queries such as `cs/web/xss`.
```
[the remark this answers is not in the document]
Comment (joefarebrother) [reply csharp/ql/src/change-notes/2023-10-24-xss-flow-steps.md:4]: What term should be used instead?

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll (hunk 12 above)
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:84]: Perhaps. Do we want to add a mechanism for defining additional value steps for all configurations? And are there likely to be many cases where this matters for precision?

File: csharp/ql/src/change-notes/2023-10-24-xss-flow-steps.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (tamasvajk) [reply csharp/ql/src/change-notes/2023-10-24-xss-flow-steps.md:4]: How about "Razor view", "Razor view file", or simply "view file"? The latter seems to be used in the [docs](https://learn.microsoft.com/en-us/aspnet/core/mvc/views/overview).

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll (hunk 12 above)
Comment (hvitved) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:84]: It matters, especially if data is stored inside a field/property of the model (which I guess is often the case?), since only value-preserving steps are allowed when tracking data stored in a field/property. However, we need to be sure that the step is indeed value-preserving.

We could add a mechanism for adding additional value steps, just like you added one for additional taint steps.

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll (hunk 8 above)
Comment (sidshank) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:121]: @michaelnebel Does [this response](https://github.com/github/codeql/pull/14343#discussion_r1374373115) from @joefarebrother on an earlier comment also address your question here?

Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:121]: Yes, it does. The link in the other response lists the view discovery as given by the order in this predicate.

File: csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/gen_files.py (hunk 1 above)
Comment (michaelnebel) [reply csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/gen_files.py:1]: It appears that one need to provide the full path to script when executing it as it is needed for the working dir.
The following works for me:
```
python3 /Users/michaelnebel/Work/semmle-code/ql/csharp/ql/test/query-tests/Security\ Features/CWE-079/XSSRazorPages/gen_files.py
``
Where as (when standing in the directory with the script) doesn't work.
```
python3 gen_files.py
```
Maybe it would be worth adding a couple of print statements to the script to get an indication where files are written and also how to invoke the script (maybe `/ql/src/Stubs/make_stubs_nuget.py` could provide some inspiration)

File: csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/gen_files.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+import sys
+import os
+
+work_dir = os.path.dirname(sys.argv[0])
```
Comment (michaelnebel) [inline csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/gen_files.py:6]:
```suggestion
work_dir = os.path.abspath(os.path.dirname(sys.argv[0]))
```

Comment (michaelnebel) [reply csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/gen_files.py:6]: With this change it was possible for me to invoke the script by `python3 gen_files.py`.

File: csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/Controllers/TestController.cs [hunk 15]
```diff
@@ -0,0 +1,149 @@
+namespace test;
+
+using System.Net;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.RazorPages;
```
[the remark this answers is not in the document]
Comment (joefarebrother) [reply csharp/ql/test/query-tests/Security Features/CWE-079/XSSRazorPages/Controllers/TestController.cs]: It was not.

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+    attr.getFile() = this.getFile() and
+    attr.getType()
+        .hasQualifiedName("Microsoft.AspNetCore.Razor.Hosting", "RazorCompiledItemAttribute")
+  }
+
+  /**
+   * Gets the filepath of the source file that this class was generated from,
+   * relative to the application root.
+   */
+  string getSourceFilepath() { result = attr.getArgument(2).(StringLiteral).getValue() }
```
[the remark this answers is not in the document]
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:84]: Do you have an example of a database for which the generated files contain absolute paths?

[the remark this answers is not in the document]
Comment (tamasvajk) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:84]: Any standalone extracted web project will have that. So running the following produces one:
```
dotnet new mvc --name abc
CODEQL_EXTRACTOR_CSHARP_STANDALONE_EXTRACT_WEB_VIEWS=true codeql database create DB_standalone --language=csharp --source-root=abc -Obuildless=true
```

The DB contains `A/DB_standalone/[...]/DB_standalone/working/razor/907F0B60B9944B50BE8465FB6805A19A/Microsoft.NET.Sdk.Razor.SourceGenerators/Microsoft.NET.Sdk.Razor.SourceGenerators.RazorSourceGenerator/[...]_abc_Views_Home_Index_cshtml.g.cs` which has the attribute with the absolute path:
```C#
[assembly: global::Microsoft.AspNetCore.Razor.Hosting.RazorCompiledItemAttribute(typeof(AspNetCoreGeneratedDocument.[...]_abc_Views_Home_Index), @"mvc.1.0.view", @"[...]/abc/Views/Home/Index.cshtml")]
```

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+class RazorPageClass extends Class {
+  AssemblyAttribute attr;
+
+  RazorPageClass() {
+    this.getBaseClass()
+        .getUnboundDeclaration()
+        .hasQualifiedName("Microsoft.AspNetCore.Mvc.Razor", "RazorPage<>") and
+    attr.getFile() = this.getFile() and
+    attr.getType()
+        .hasQualifiedName("Microsoft.AspNetCore.Razor.Hosting", "RazorCompiledItemAttribute")
```
[the remark this answers is not in the document]
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:79]: I have tried that at some point, and it seemed to not work due to that attribute not appearing in (non-standalone) DBs.

[the remark this answers is not in the document]
Comment (tamasvajk) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:79]: That's odd and unfortunate. I guess it could also depend on the dotnet version being used. I'm somewhat worried that https://github.com/github/codeql/pull/14792 doesn't fix the assembly attribute extraction, and then we wouldn't find the `RazorCompiledItemAttribute` attributes.

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll (hunk 11 above)
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:43]: According to [this](https://stackoverflow.com/questions/33678831/create-action-method-on-controller-with-open-generic-type) SO question, action methods cannot be generic.

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll (hunk 17 above)
[the remark this answers is not in the document]
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:79]: In my integration tests, the `RazorCompiledItemAttribute` attributes are successfully found.

**Review by tamasvajk — APPROVED** [review]
LGTM. Let's wait for feedback from the others too.

Have you checked the new alerts in `tobyash86/WebGoat.NET`, are those true positive new findings?

**Comment by joefarebrother** [comment]
Yes, the new results in `tobyash86/WebGoat.NET` look like TPs.

**Review by michaelnebel — COMMENTED** [review]
LGTM! 👍

**Comment by joefarebrother** [comment]
Should I worry about the failing integration tests? They don't appear to be related to these changes, but I've tied rerunning them a few times.

**Comment by michaelnebel** [comment]
> Should I worry about the failing integration tests? They don't appear to be related to these changes, but I've tied rerunning them a few times.

Yeah, we need them to pass.
Try and rebase again - they were just fixed a couple of days ago.
